clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,323 @@
+		created, ipAssigned, reachable := 0, 0, 0
+		select {
+		case <-time.After(s.waitTimeout):
+			return fmt.Errorf("waiting for services timeout")
+		case <-time.After(defaultCheckInterval):
+			func() {
+				reachable = s.creationTimes.Count(reachabilityPhase)
+				ipAssigned = s.creationTimes.Count(ipAssigningPhase)
+				created = s.creationTimes.Count(creatingPhase)
+
+			}()
+			klog.Infof("%s: %d created, %d ipAssigned, %d reachable", s, created, ipAssigned, reachable)
+			if created == reachable {
+				return nil
+			}
+		}
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering service created latency measurement...", s)
+	if !s.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", s)
+	}
+
+	serviceCreationLatency := s.creationTimes.CalculateTransitionsLatency(map[string]measurementutil.Transition{
+		"create_to_assigned": {
+			From: creatingPhase,
+			To:   ipAssigningPhase,
+		},
+		"assigned_to_available": {
+			From: ipAssigningPhase,
+			To:   reachabilityPhase,
+		},
+		"create_to_available": {
+			From: creatingPhase,
+			To:   reachabilityPhase,
+		},
+	})
+
+	content, err := util.PrettyPrintJSON(measurementutil.LatencyMapToPerfData(serviceCreationLatency))
+	if err != nil {
+		return nil, err
+	}
+	summary := measurement.CreateSummary(serviceCreationLatencyName, "json", content)
+	return []measurement.Summary{summary}, nil
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	if isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status); isEqual {
+		return
+	}
+
+	// TODO(#680): Make it thread-safe.
+	if !s.isRunning {
+		return
+	}
+	if newObj == nil {
+		if err := s.deleteObject(oldService); err != nil {
+			klog.Errorf("%s: delete checker error: %v", s, err)
+		}
+		return
+	}
+	if err := s.updateObject(newService); err != nil {
+		klog.Errorf("%s: create checker error: %v", s, err)
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) deleteObject(svc *corev1.Service) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(svc)
+	if err != nil {
+		return fmt.Errorf("meta key created error: %v", err)
+	}
+	s.pingCheckers.DeleteAndStop(key)
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) updateObject(svc *corev1.Service) error {
+	// This measurement only works for services with LoadBalancer type.
+	// Service of this type has to have port specified.
+	if len(svc.Spec.Ports) < 1 {
```

Comment:
BTW - when I know think about it, we can simply check svc.Spec.Type instead - it will be consistent with what we're saying we support.

Reply:
Done.